native/dll/DShowCapture-2/GraphManager.cpp

```diff
@@ -362,8 +401,13 @@ JNIEXPORT void JNICALL Java_sage_DShowCaptureDevice_teardownGraph0
 		} else
 		{
 			TeardownBDAGraph( env, pCapInfo->captures[1] );
+		} else
+		{
+			TeardownBDAGraph( env, pCapInfo->captures[2] );
+		} else
+		{
+			TeardownBDAGraph( env, pCapInfo->captures[3] );
 		}
-
```

Comment:
I need some help with this. I'm assuming this is where I need to do a loop as else are obviously not right here

Reply:
Does it need to something like this? Not sure if I should be testing against just num 4 or something else
```
		if ( pCapInfo->captures[0]->dwBDAType != pCapInfo->dwBDAType )
		for (int i = 0; i<4; i++)
		{
			TeardownBDAGraph( env, pCapInfo->captures[i] );
		} 

```

Reply:
For the teardown, put the compare inside the loop like this:

` 
		
	int i;

	for (i = 0; i<4; i++)
	{
		if ( pCapInfo->captures[i]->dwBDAType != pCapInfo->dwBDAType )
			TeardownBDAGraph( env, pCapInfo->captures[i] );
	}
`

Also needed [i] instead of [0] in the 'if'.  This will tear down all except the current one.

.
.
.

To loop the TuningMode assignment stuff consider this:

`

		DeviceCaps = deviceCaps & ~BDA_CAPTURE_ALL;
		
		int i,j;

		for (i=0; i<4; i++)
		{
			strncpy( TuningMode, GetBDAType( deviceCaps, i ), sizeof(TuningMode) );
			slog((env, "Create first DshowCapture %s.\r\n", TuningMode ));
			rvs[i] = CreateDshowCaptureInfo( env, videoCaptureFilterName, capDevNum, deviceCaps, tvType, 
				                                 TuningMode, Country );
		}


		for (i=0; i<4; i++)
		{
			if ( rvs[i] )
			{
				rvs[i]->captureNum = 4;
				for (j=0; j<4; j++)
					rvs[i]->captures[j] = rvs[j];
			}
		}

		rv = new DShowCaptureInfo;
		memset( rv, 0, sizeof(rv) );

		for (i=0; i<4; i++)
		{
			if ( rvs[i] )
			{
				memcpy( rv, rvs[i], sizeof(DShowCaptureInfo) );
				break;
			}
		}

		if (i==4)		// no valid rvs
			return 0;

		setChannelDev( (CHANNEL_DATA*)rv->channel, (void*)rv );
		return (jlong)rv;

`
.
.
.
Also don't forget to fix the 'deletes' at the bottom of 'Java_sage_DShowCaptureDevice_teardownGraph0':

`

	int i;
	for (i=0; i<4; i++)
	{
		if ( pCapInfo->captures[i] )
			delete pCapInfo->captures[i];
	}
`
.
.
.
Lastly, did you remember to change this line in DShowCapture.h?

`#define CAPTURE_NUM 4`
That sets the size for the 'captures' data structures.

Reply:
Thanks wnjj. I really appreciate you help. It does emphasize to me how little I know about this though.:)
I've made the changes but now SageTV crashes on startup when it gets to the tuningmode assignment code, so just trying to debug that.